src/lib.rs

```diff
@@ -111,18 +165,249 @@ macro_rules! nonmax {
+                unsafe { Self::new_unchecked(small.into()) }
+            }
         }
     };
 }
 
-nonmax!(NonMaxI8, NonZeroI8, i8);
-nonmax!(NonMaxI16, NonZeroI16, i16);
-nonmax!(NonMaxI32, NonZeroI32, i32);
-nonmax!(NonMaxI64, NonZeroI64, i64);
-nonmax!(NonMaxIsize, NonZeroIsize, isize);
-
-nonmax!(NonMaxU8, NonZeroU8, u8);
-nonmax!(NonMaxU16, NonZeroU16, u16);
-nonmax!(NonMaxU32, NonZeroU32, u32);
-nonmax!(NonMaxU64, NonZeroU64, u64);
-nonmax!(NonMaxUsize, NonZeroUsize, usize);
+// Unsigned -> Non-max Unsigned
+impl_smaller_from!(u8, NonMaxU16);
+impl_smaller_from!(u8, NonMaxU32);
+impl_smaller_from!(u8, NonMaxU64);
+impl_smaller_from!(u8, NonMaxU128);
+impl_smaller_from!(u8, NonMaxUsize);
+impl_smaller_from!(u16, NonMaxU32);
+impl_smaller_from!(u16, NonMaxU64);
+impl_smaller_from!(u16, NonMaxU128);
+impl_smaller_from!(u16, NonMaxUsize);
+impl_smaller_from!(u32, NonMaxU64);
+impl_smaller_from!(u32, NonMaxU128);
+impl_smaller_from!(u64, NonMaxU128);
+
+// Signed -> Non-max Signed
+impl_smaller_from!(i8, NonMaxI16);
+impl_smaller_from!(i8, NonMaxI32);
+impl_smaller_from!(i8, NonMaxI64);
+impl_smaller_from!(i8, NonMaxI128);
+impl_smaller_from!(i8, NonMaxIsize);
+impl_smaller_from!(i16, NonMaxI32);
+impl_smaller_from!(i16, NonMaxI64);
+impl_smaller_from!(i16, NonMaxI128);
+impl_smaller_from!(i16, NonMaxIsize);
+impl_smaller_from!(i32, NonMaxI64);
+impl_smaller_from!(i32, NonMaxI128);
+impl_smaller_from!(i64, NonMaxI128);
+
+// Unsigned -> Non-max Signed
+impl_smaller_from!(u8, NonMaxI16);
+impl_smaller_from!(u8, NonMaxI32);
+impl_smaller_from!(u8, NonMaxI64);
+impl_smaller_from!(u8, NonMaxI128);
+impl_smaller_from!(u8, NonMaxIsize);
+impl_smaller_from!(u16, NonMaxI32);
+impl_smaller_from!(u16, NonMaxI64);
+impl_smaller_from!(u16, NonMaxI128);
+impl_smaller_from!(u32, NonMaxI64);
+impl_smaller_from!(u32, NonMaxI128);
+impl_smaller_from!(u64, NonMaxI128);
+
+fn parse_int_error_overflow() -> std::num::ParseIntError {
+    use std::str::FromStr;
+    u8::from_str("999").unwrap_err()
+}
+
+fn try_from_int_error_overflow() -> std::num::TryFromIntError {
+    use std::convert::TryFrom;
+    u8::try_from(999u32).unwrap_err()
+}
+
+#[cfg(test)]
+mod ops {
+    use super::*;
+
+    #[test]
+    fn bitand_unsigned() {
+        for left in 0..=255 {
+            let nmleft = NonMaxU8::new(left);
+            for right in 0..=255 {
+                let nmright = NonMaxU8::new(right);
+                let vanilla = left & right;
+
+                if let (Some(nmleft), Some(nmright)) = (nmleft, nmright) {
+                    assert_eq!(vanilla, (nmleft & nmright).get());
+                }
+                if let Some(nmleft) = nmleft {
+                    assert_eq!(vanilla, (nmleft & right).get());
+                }
+                if let Some(nmright) = nmright {
+                    assert_eq!(vanilla, (left & nmright).get());
+                }
+            }
+        }
+    }
```

Comment:
These tests are very cool.

Can we use the std constants for the value ranges on these types? While I'm sure everyone knows the range of u8 and i8, using `u8::MIN` and `u8::MAX` might be a little more clear.